### PR TITLE
Ticket5084 multiple consoles provide warning

### PR DESCRIPTION
**Pull request by RaiBishal**
### Description of work

Warn If I open multiple scripting consoles so that I don't inadvertently have two (or more) consoles running scripts.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/5084

### Acceptance criteria

1. On opening a second scripting console, I am warned that I already have an open console.
1. IBEX provides me with a choice:
   1. Proceed to open a second console.
   1. Cancel the opening of a second console.
1. `Pin console` icon is not shown
1. "select python type" dialog does not appear when creating a new console.

### Unit tests

*Give an overview of unit tests you have added or modified, if applicable. The aim is provide information to help the reviewer*

### System tests

*Mention any automated tests or manual tests that you have added or modified, if applicable. The aim is provide information to help the reviewer*

### Documentation
*Highlight and provide a link to any additions or changes to the documentation, if applicable. The aim is provide information to help the reviewer*

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by RaiBishal** [comment]
> Also, the original ticket said:
> 
> > It would also be helpful if the console title could show
> > if a script is running
> > the name of the running script (if any)
> 
> Which this hasn't done, I think it will be really hard to do so happy for you not to do it as part of this ticket but can you spin it into a new one?

I was thinking if we could split this PyDev work such that one ticket is focusing on console dependent acceptance criteria (this one possibly https://github.com/ISISComputingGroup/IBEX/issues/1154) and maybe this ticket focusing on console independent acceptance criteria.
Features such as showing the name of the script and adding the title are only shown when the console is open in scripting perspective(console dependent). The reason behind is that (**big**)if we have to write our own PyDev console to meet other acceptance criteria we could still use the bits we have done on this PR. 
If this sounds an overkill I can add other requested features here too.

**Comment by DominicOram** [comment]
I agree that we should have a separate ticket for the running script name thing. I think just make a completely new one rather than reuse the old one though.

Do you think having a warning on the kill console (as mentioned in https://github.com/ISISComputingGroup/IBEX/issues/1154) will also involve having to write our own console? I haven't thought about it long enough to be sure.

**Comment by RaiBishal** [comment]
`kill console` icon creation is tightly coupled with console creation. If we could find a way to delete the `kill console` icon and add our own button(which give out a warning) then we can avoid having to write our own console.

**Comment by DominicOram** [comment]
Also, can we have a squish test for this?

File: base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/GeniePythonConsoleFactory.java [hunk 1]
```diff
@@ -97,16 +101,36 @@ public void configureAndCreateConsole(String additionalInitialCommands, boolean
...
-			LOG.error(e);
-		}
+		
+		Display.getDefault().syncExec(new Runnable() {
+		    public void run() {
+		    	IConsoleManager manager = ConsolePlugin.getDefault().getConsoleManager();
+		    	boolean continueWithNewConsole = true;
+		    	if (manager.getConsoles().length >= 1 && Consoles.showDialog) {
+		    		continueWithNewConsole = MessageDialog.openQuestion(Display.getDefault().getActiveShell(), "Duplicate console",
+			    			"Scripting console already Exists\nDo you still want to proceed with a new one?");
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/GeniePythonConsoleFactory.java]: Added above information.

File: base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [DEFAULT]
-uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=uk.ac.stfc.isis.ibex.client.e4.product.perspective.scriptGenerator,uk.ac.stfc.isis.ibex.client.e4.product.perspective.reflectometry
+#uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=uk.ac.stfc.isis.ibex.client.e4.product.perspective.scriptGenerator,uk.ac.stfc.isis.ibex.client.e4.product.perspective.reflectometry
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini:2]: Reverted.

File: base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/Consoles.java [hunk 3]
```diff
@@ -123,13 +136,45 @@ public void handleEvent(Event event) {
...
+	}
+	
+	private void editToolbarItems() {
+		Display.getDefault().asyncExec(new Runnable() {
+			@Override
+			public void run() {
+				IWorkbenchWindow window = PlatformUI.getWorkbench().getActiveWorkbenchWindow();
+				IWorkbenchPage page = window.getActivePage();
+				IViewPart part = page.findView(IConsoleConstants.ID_CONSOLE_VIEW);
+				ConsoleView view = (ConsoleView)part;
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/Consoles.java]: Done.

File: base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/Consoles.java [hunk 4]
```diff
@@ -123,13 +136,45 @@ public void handleEvent(Event event) {
...
+					IContributionItem [] items= tbm.getItems();
+					// Add our "Open Console" action
+					GenieOpenConsoleAction openConsoleAction = new GenieOpenConsoleAction();
+					tbm.insertBefore(WARNING_CHECKBOX_ID, openConsoleAction);
+					
+					// Console view icons not required
+					List<ActionContributionItem> itemsToRemove = Arrays.stream(items).filter(item->(item instanceof ActionContributionItem))
+ 					.map(item->(ActionContributionItem)item)
+					.filter(item-> item.getAction().toString().contains("PinConsole")||
+							item.getAction() instanceof OpenConsoleAction)
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/Consoles.java:167]: Done.

File: base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/GenieOpenConsoleAction.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+				if (e2.isNewConsoleExtenson()) {
+					return 1;
+				}
+				String first = e1.getLabel();
+				String second = e2.getLabel();
+				return first.compareTo(second);
+			}
+		});
+		return factoryExtensions;
+	}
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/GenieOpenConsoleAction.java]: Indeed. Removed this bit.

File: base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/GenieOpenConsoleAction.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+					label = "PyDev Console";
+					image = pyDevExt.getImageDescriptor();
+				}
+				addActionToMenu(fMenu, new ConsoleFactoryAction(label, image, extension), accel);
+				accel++;
+				if (extension.isNewConsoleExtenson()) {
+					new Separator("new").fill(fMenu, -1); //$NON-NLS-1$
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (RaiBishal) [reply base/uk.ac.stfc.isis.ibex.ui.scripting/src/uk/ac/stfc/isis/ibex/ui/scripting/GenieOpenConsoleAction.java]: refactored.

**Comment by RaiBishal** [comment]
Squish Test PR:
https://github.com/ISISComputingGroup/System_Tests_UI_E4/pull/45

File: base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (DominicOram) [reply base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini:2]: This is still here?

**Pull request merged**